app/controllers/graphql_controller.rb

```diff
@@ -0,0 +1,111 @@
+# frozen_string_literal: true
+
+class GraphqlController < ApplicationController
+  # Note 22/02/10 - Nimmo
+  # This controller is short but important.
+  # Parses an incoming request header to build a Context hash (authentication)
+  # and forwards that to the graphql Query, w/ Variables and Operation Name.
+  # Possible operation names: Query, Mutation, Union etc
+
+  disable_filters
+
+  def execute
+    variables = prepare_variables(params[:variables])
+    query = params[:query]
+    operation_name = params[:operationName]
+    context = {
+      current_user: current_user,
+      in_admin_mode?: in_admin_mode?
+    }
+    result = MushroomObserverSchema.execute(
+      query,
+      variables: variables,
+      context: context,
+      operation_name: operation_name
+    )
+
+    render(json: result)
+  rescue StandardError => e
+    raise(e) unless Rails.env.development?
+
+    handle_error_in_development(e)
+  end
+
+  private
+
+  # https://www.howtographql.com/graphql-ruby/4-authentication/
+  # Decrypt the current user from token stored in the header (not the session)
+  # Note: There may not be a token!
+  def current_user
+    return unless token.is_a?(Hash)
```

Comment:
I would consider making these ... oh shoot there's a term for it, but I'm forgetting the lingo...  ah! "memo"... I would consider memoizing this method and a few of the others below.  All the ones you can reasonably expect to need multiple times for a given query and which involve significant processing.  Basically, fix it to do all this processing once, and then cache the result so it doesn't have to do it again on subsequent calls.  (Things that can return nil may require some tweaking.)

Reply:
This works for me.  The main thing to avoid having to do every time is User.safe_find.  Thanks!